firefox-ios/Client/Frontend/Browser/QRCodeViewController.swift

```diff
@@ -335,14 +355,45 @@ extension QRCodeViewController: AVCaptureMetadataOutputObjectsDelegate {
                let text = metaData.stringValue {
                 // Open QR codes only when they are recognized as webpages, otherwise open as text
                 if let url = URIFixup.getURL(text), url.isWebPage() {
-                    qrCodeDelegate.didScanQRCodeWithURL(url)
+                    let shouldPrompt = qrCodeDelegate.qrCodeScanningPermissionLevel != .allowURLsWithoutPrompt
+
+                    if shouldPrompt {
+                        // Prompt users before opening scanned URL. Show shortened host (if possible) as the preview.
+                        let websiteHost: String?
+                        if #available(iOS 16.0, *) { websiteHost = url.host() } else { websiteHost = url.host }
+                        let presentedURL = websiteHost ?? text
+
+                        let confirmationAlert = UIAlertController(
+                            title: .ScanQRCodeConfirmOpenURLMessage,
+                            message: presentedURL,
+                            preferredStyle: .alert)
+
+                        let cancelAction = UIAlertAction(title: .ScanQRCodeURLPromptDenyButton,
+                                                         style: .cancel,
+                                                         handler: { _ in
+                            openScannedQRCodeURL(url)
+                            cleanUpAndRemoveQRCodeScanner()
+                        })
+                        let openURLAction = UIAlertAction(title: .ScanQRCodeURLPromptAllowButton,
+                                                          style: .default,
+                                                          handler: { [weak self] _ in
+                            self?.qrCodeDelegate?.didScanQRCodeWithURL(url)
+                            cleanUpAndRemoveQRCodeScanner()
+                        })
+
+                        confirmationAlert.addAction(cancelAction)
+                        confirmationAlert.addAction(openURLAction)
+
+                        self.present(confirmationAlert, animated: true)
+                        return
+                    } else {
+                        openScannedQRCodeURL(url)
+                    }
```

Comment:
nit: do we want to add this into its own private method so this method is more readable?

Reply:
Yeah I was also considering that, I kept it within the existing func because it didn't seem like that much code and several things would need to get broken out otherwise. But I'm happy to refactor this to split it up if you have a strong preference, lmk.

Reply:
no strong preference! thanks for considering this.